packages/grid/x-data-grid/src/hooks/features/rows/useGridRowsMeta.ts

```diff
@@ -32,41 +32,77 @@ export const useGridRowsMeta = (
+    };
   }>({});
-  const rowHeight = useGridSelector(apiRef, gridDensityRowHeightSelector);
+
+  // Inspired by https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/utils/CellSizeAndPositionManager.js
+  const lastMeasuredRowIndex = React.useRef(0);
+  const hasRowWithAutoHeight = React.useRef(false);
+  const rowHeightFromDensity = useGridSelector(apiRef, gridDensityRowHeightSelector);
   const filterState = useGridSelector(apiRef, gridFilterStateSelector);
   const paginationState = useGridSelector(apiRef, gridPaginationSelector);
   const sortingState = useGridSelector(apiRef, gridSortingStateSelector);
   const currentPage = useGridVisibleRows(apiRef, props);
 
   const hydrateRowsMeta = React.useCallback(() => {
+    hasRowWithAutoHeight.current = false;
+
     apiRef.current.setState((state) => {
-      const positions: number[] = [];
       const densityFactor = gridDensityFactorSelector(state, apiRef.current.instanceId);
-      const currentRowHeight = gridDensityRowHeightSelector(state, apiRef.current.instanceId);
-      const currentPageTotalHeight = currentPage.rows.reduce((acc: number, row) => {
+
+      const positions: number[] = [];
+      const currentPageTotalHeight = currentPage.rows.reduce((acc, row) => {
         positions.push(acc);
-        let baseRowHeight: number;
 
-        const isResized =
-          (rowsHeightLookup.current[row.id] && rowsHeightLookup.current[row.id].isResized) || false;
+        if (!rowsHeightLookup.current[row.id]) {
+          rowsHeightLookup.current[row.id] = {
+            sizes: { base: rowHeightFromDensity },
+            isResized: false,
+            autoHeight: false,
+            needsFirstMeasurement: true, // Assume all rows will need to be measured by default
+          };
+        }
 
-        if (isResized) {
-          // do not recalculate resized row height and use the value from the lookup
-          baseRowHeight = rowsHeightLookup.current[row.id].value;
-        } else {
-          baseRowHeight = currentRowHeight;
+        const { isResized, needsFirstMeasurement, sizes } = rowsHeightLookup.current[row.id];
+        let baseRowHeight = rowHeightFromDensity;
+        const existingBaseRowHeight = sizes.base;
 
-          if (getRowHeight) {
+        if (isResized) {
+          // Do not recalculate resized row height and use the value from the lookup
+          baseRowHeight = existingBaseRowHeight;
+        } else if (getRowHeightProp) {
+          const rowHeightFromUser = getRowHeightProp({ ...row, densityFactor });
+
+          if (rowHeightFromUser === 'auto') {
+            const estimatedRowHeight = getEstimatedRowHeight
+              ? getEstimatedRowHeight({ ...row, densityFactor })
+              : rowHeightFromDensity;
+
+            // If the row was not measured yet use the estimated row height
+            baseRowHeight = needsFirstMeasurement
+              ? estimatedRowHeight ?? rowHeightFromDensity
+              : existingBaseRowHeight;
```

Comment:
```suggestion
            if (needsFirstMeasurement) {
              const estimatedRowHeight = getEstimatedRowHeight
                ? getEstimatedRowHeight({ ...row, densityFactor })
                : rowHeightFromDensity;
  
              baseRowHeight = estimatedRowHeight ?? rowHeightFromDensity
            } else {
              baseRowHeight = existingBaseRowHeight;
            }
```

Just a small tweak so that the `getEstimatedRowHeight` is only called when there is a need to obtain the estimated row height